### PR TITLE
[CHAT-2592] ShadowBan & Pin methods

### PR DESCRIPTION
**Pull request by gumuz**
Add ShadowBan, RemoveShadowBan, PinMessage & UnPinMessage methods.

### CODE REVIEW

File: channel.go [hunk 1]
```diff
@@ -437,6 +437,32 @@ func (ch *Channel) UnBanUser(targetID string, options map[string]string) error {
 	return ch.client.UnBanUser(targetID, options)
 }
 
+// ShadowBan bans target userID from this channel
+// userID: user who shadow bans target.
+// options: additional shadow ban options, ie {"timeout": 3600, "reason": "offensive language is not allowed here"}.
+func (ch *Channel) ShadowBan(targetID, userID string, options map[string]interface{}) error {
```
Comment (mahboubii) [inline channel.go:443]: difference between target and userID is not clear

File: channel.go [hunk 2]
```diff
@@ -437,6 +437,32 @@ func (ch *Channel) UnBanUser(targetID string, options map[string]string) error {
...
+	}
+
+	options["type"] = ch.Type
+	options["id"] = ch.ID
+
+	return ch.client.ShadowBan(targetID, userID, options)
+}
+
+// RemoveShadowBan removes the ban for target user ID on this channel.
+func (ch *Channel) RemoveShadowBan(targetID string, options map[string]string) error {
```
Comment (mahboubii) [inline channel.go:455]: instead of targetID maybe better to just use userID?

File: message.go [hunk 3]
```diff
@@ -269,6 +273,60 @@ func (c *Client) UpdateMessage(msg *Message, msgID string) (*Message, error) {
 	return resp.Message, nil
 }
 
+// PartialUpdateMessage partially updates message with given msgID.
+func (c *Client) PartialUpdateMessage(msgID, userID string, updates PartialUpdate) (*Message, error) {
```
Comment (mahboubii) [inline message.go:277]: we can make it clear here that userID is optional field

File: message.go [hunk 4]
```diff
@@ -269,6 +273,60 @@ func (c *Client) UpdateMessage(msg *Message, msgID string) (*Message, error) {
...
+	var resp messageResponse
+
+	p := path.Join("messages", url.PathEscape(msgID))
+
+	options := map[string]interface{}{
+		"Set":   updates.Set,
+		"Unset": updates.Unset,
+		"User": map[string]string{
+			"id": userID,
+		},
```
Comment (mahboubii) [inline message.go:294]: would API throw for empty/nil `user.id`?

File: message.go (hunk 3 above)
Comment (mahboubii) [reply message.go:277]: PartialUpdate also accepts `options{skip_enrich_url: bool}`

File: channel.go [hunk 5]
```diff
@@ -437,6 +437,32 @@ func (ch *Channel) UnBanUser(targetID string, options map[string]string) error {
 	return ch.client.UnBanUser(targetID, options)
 }
 
+// ShadowBan shadow bans userID from this channel
+// bannedByID: user who shadow bans userID.
+// options: additional shadow ban options, ie {"timeout": 3600, "reason": "offensive language is not allowed here"}.
+func (ch *Channel) ShadowBan(userID, bannedByID string, options map[string]interface{}) error {
```
Comment (mahboubii) [inline channel.go:443]: should we type the options since it only accepts certain fields?

File: channel.go [hunk 6]
```diff
@@ -437,6 +437,32 @@ func (ch *Channel) UnBanUser(targetID string, options map[string]string) error {
...
+	}
+
+	options["type"] = ch.Type
+	options["id"] = ch.ID
+
+	return ch.client.ShadowBan(userID, bannedByID, options)
+}
+
+// RemoveShadowBan removes the ban for target user ID on this channel.
+func (ch *Channel) RemoveShadowBan(userID string, options map[string]string) error {
```
Comment (mahboubii) [inline channel.go:455]: same here, not clear what options we are accepting

File: channel.go (hunk 5 above)
Comment (gumuz) [reply channel.go:443]: not sure, makes it less flexible. we use interface maps everywhere

Comment (gumuz) [reply channel.go:443]: @yaziine what do you think?

File: channel.go [hunk 7]
```diff
@@ -437,6 +437,32 @@ func (ch *Channel) UnBanUser(targetID string, options map[string]string) error {
...
+		options = map[string]interface{}{}
+	}
+
+	options["type"] = ch.Type
+	options["id"] = ch.ID
+
+	return ch.client.ShadowBan(userID, bannedByID, options)
+}
+
+// RemoveShadowBan removes the ban for target user ID on this channel.
```
Comment (yaziine) [inline channel.go:454]:
```suggestion
// RemoveShadowBan removes the shadow ban for target user ID on this channel.
```

File: channel.go (hunk 6 above)
Comment (yaziine) [reply channel.go:455]: following the discussion above, i think we should make it crystal clear which options are acceptable in the go doc.

But here for `RemoveShadowBan`, i am not sure it is possible to pass any extra options.

File: message.go [hunk 8]
```diff
@@ -269,6 +273,70 @@ func (c *Client) UpdateMessage(msg *Message, msgID string) (*Message, error) {
 	return resp.Message, nil
 }
 
+// PartialUpdateMessage partially updates message with given msgID.
+func (c *Client) PartialUpdateMessage(msgID string, updates PartialUpdate, options map[string]interface{}) (*Message, error) {
+	switch {
+	case len(updates.Set) == 0 && len(updates.Unset) == 0:
+		return nil, errors.New("updates is empty")
```
Comment (yaziine) [inline message.go:280]: `errors.New("updates are empty")`  or `errors.New("updates parameter is empty")`

File: message.go [hunk 9]
```diff
@@ -269,6 +273,70 @@ func (c *Client) UpdateMessage(msg *Message, msgID string) (*Message, error) {
 	return resp.Message, nil
 }
 
+// PartialUpdateMessage partially updates message with given msgID.
+func (c *Client) PartialUpdateMessage(msgID string, updates PartialUpdate, options map[string]interface{}) (*Message, error) {
```
Comment (yaziine) [inline message.go:277]: Same here which options can be passed to this method?

File: message.go [hunk 10]
```diff
@@ -269,6 +273,70 @@ func (c *Client) UpdateMessage(msg *Message, msgID string) (*Message, error) {
...
+
+// PinMessage pins the message with given msgID.
+func (c *Client) PinMessage(msgID, pinnedByID string, expiration *time.Duration) (*Message, error) {
+	updates := PartialUpdate{
+		Set: map[string]interface{}{
+			"pinned": true,
+		},
+	}
+	if expiration != nil {
+		updates.Set["pin_expires"] = int(expiration.Milliseconds())
```
Comment (yaziine) [inline message.go:315]: Why do we need to cast here?

**Review by yaziine — APPROVED** [review]

**Pull request merged**